app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+                    'message' => 'Save failed',
+                    'trace' => 'Calculated checksum : '.$calculatedChecksum,
+                    'code' => 900
+                ]
+            ];
+        }
+
+        $autoSaver['currentContent'] = $post->content;
+        $autoSaver['count'] = (int)$autoSaver['count'] + 1;
+
+        $request->session()->set('autoSave'.$postId, $autoSaver);
+
+        $updatedTime = Carbon::now()->toDateTimeString();
+
+        //$history = null;
+
+        //if($autoSaver['count'] % PostHistory::AUTOSAVE_INTERVAL == 0) {
+            $history = $post->saveHistory(PostHistory::STATUS_DRAFT);
+            $updatedTime = $history->updated_at;
+        //}
+
+        if(!$post->isPublished()) {
+            $post->save();
+            $updatedTime = $post->updated_at;
+        }
+
+        $returnData = [
+            "id" => $post->id,
+            "time" => $updatedTime,
+        ];
+
+        if(env('APP_DEBUG', false)) {
+            $returnData["debug"] = [
+                "count" => $request->session()->get('autoSave'.$postId)['count'],
+                "saved" => ($history != null)
+            ];
+        }
+
+        return $returnData;
+    }
+
+    public function postStashChange(Request $request) {
+        $postId = $request->get('post_id', 0);
+
+        $post = Post::find($postId);
+
+        if(empty($post) || ($post->user_id != $this->currentUser->id)) {
+            return ['error' => 'Post not found'];
+        }
+
+        // Lock detect
+        if($post->isLock($this->currentSession)) {
+            return [
+                'id' => $post->id,
+                'status' => 'locked'
+            ];
+        }
+
+        $post->updated_at = Carbon::now()->toDateTimeString();
+        $post->save();
+
+        PostHistory::saveFromSession($request->session(), $post);
+
+        return ['id' => $post->id];
+    }
+
+    public function postBackupUri(Request $request) {
+        $postId = $request->get('post_id');
+        $metaData = $request->get('data');
+
+        $post = Post::find($postId);
+
+        if(empty($post) || ($post->user_id != $this->currentUser->id)) {
+            return ['error' => [
+                'code' => 404,
+                'message' => 'Post not found'
+            ]];
+        }
+
+        Backup::newBackup($post->id, $this->currentSession->session_key, $metaData);
+
+        return ['id' => $post->id];
+    }
+
+    public function postLock(Request $request) {
+        $postId = $request->get('post_id', 0);
+        $postIdLog = $request->get('post_id_log', 0);
+
+        $post = Post::find($postId);
+        if(empty($post) || !$post->canEditBy($this->currentUser)) {
+            $this->activityLog->setPostId($postIdLog);
+            if(empty($post)) {
+                $this->activityLog->addMeta('LockError', 'Post does not exist.');
+            } else if(!$post->canEditBy($this->currentUser)) {
+                $this->activityLog->addMeta('Lock
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS112)